src/Parsers/Access/ASTCheckGrantQuery.cpp

```diff
@@ -0,0 +1,126 @@
+        if (element.isGlobalWithParameter())
+        {
+            if (element.anyParameter())
+                settings.ostr << "*";
+            else
+                settings.ostr << backQuoteIfNeed(element.parameter);
+        }
+        else if (element.anyDatabase())
+        {
+            settings.ostr << "*.*";
+        }
+        else
+        {
+            if (!element.database.empty())
+                settings.ostr << backQuoteIfNeed(element.database) << ".";
+            if (element.anyDatabase())
+                settings.ostr << "*";
+            else
+                settings.ostr << backQuoteIfNeed(element.table);
+        }
+    }
+
+
+    void formatElementsWithoutOptions(const AccessRightsElements & elements, const IAST::FormatSettings & settings)
+    {
+        bool no_output = true;
+        for (size_t i = 0; i != elements.size(); ++i)
+        {
+            const auto & element = elements[i];
+            auto keywords = element.access_flags.toKeywords();
+            if (keywords.empty() || (!element.anyColumn() && element.columns.empty()))
+                continue;
+
+            for (const auto & keyword : keywords)
+            {
+                if (!std::exchange(no_output, false))
+                    settings.ostr << ", ";
+
+                settings.ostr << (settings.hilite ? IAST::hilite_keyword : "") << keyword << (settings.hilite ? IAST::hilite_none : "");
+                if (!element.anyColumn())
+                    formatColumnNames(element.columns, settings);
+            }
+
+            bool next_element_on_same_db_and_table = false;
+            if (i != elements.size() - 1)
+            {
+                const auto & next_element = elements[i + 1];
+                if (element.sameDatabaseAndTableAndParameter(next_element))
+                {
+                    next_element_on_same_db_and_table = true;
+                }
+            }
+
+            if (!next_element_on_same_db_and_table)
+            {
+                settings.ostr << " ";
+                formatONClause(element, settings);
+            }
+        }
+
+        if (no_output)
+            settings.ostr << (settings.hilite ? IAST::hilite_keyword : "") << "USAGE ON " << (settings.hilite ? IAST::hilite_none : "") << "*.*";
+    }
+
+}
+
+
+String ASTCheckGrantQuery::getID(char) const
+{
+    return "CheckGrantQuery";
+}
+
+
+ASTPtr ASTCheckGrantQuery::clone() const
+{
+    auto res = std::make_shared<ASTCheckGrantQuery>(*this);
+
+    return res;
+}
+
+
+void ASTCheckGrantQuery::formatImpl(const FormatSettings & settings, FormatState &, FormatStateStacked) const
+{
+    settings.ostr << (settings.hilite ? IAST::hilite_keyword : "") << "CHECK GRANT"
+                  << (settings.hilite ? IAST::hilite_none : "");
+
+    settings.ostr << " ";
+
+    formatElementsWithoutOptions(access_rights_elements, settings);
```

Comment:
Why can't you use
```
settings.ostr << access_rights_element.toString();
```
here instead of calling this function `formatElementsWithoutOptions()` which looks like a lot of copy-paste?

Reply:
> Why can't you use
> 
> ```
> settings.ostr << access_rights_element.toString();
> ```
> 
> here instead of calling this function `formatElementsWithoutOptions()` which looks like a lot of copy-paste?

The function does some format works in formatFunction(). Suppose I use ```settings.ostr << access_rights_element.toString();``` here, such query's format will be like (which is created by Parser):
```
:) check grant select(col,col2) on tb, alter on tb2

CHECK GRANT SELECT(col, col2) ON tb, GRANT ALTER TABLE, ALTER VIEW ON tb2

Query id: 5f3927b0-8d99-4db2-a702-67e53955476c

   ┌─result─┐
1. │      1 │
   └────────┘

1 row in set. Elapsed: 0.001 sec. 
```
However, the correct response should be just like ```GRANT``` keyword:
```

:) check grant select(col,col2) on tb, alter on tb2

CHECK GRANT SELECT(col, col2) ON tb, ALTER TABLE, ALTER VIEW ON tb2

Query id: b48aabc2-2b6e-45ed-a36a-7802a00e3852

   ┌─result─┐
1. │      1 │
   └────────┘

1 row in set. Elapsed: 0.001 sec. 
```

I will pollish the formatFunction here.

Reply:
The main problem with your function is that it looks like a lot of copy-pasted copy which isn't good. Perhaps you could get rid of at least `formatColumnNames` & `formatONClause`?